cpp/serve/radix_tree.h

```diff
@@ -0,0 +1,417 @@
+ private:
+  /*! \brief The number of nodes in sequence ID node pool. */
+  size_t num_nodes;
+  /*! \brief The sequence ID node pool. */
+  std::vector<SequenceIDNode*> nodes;
+  /*! \brief The indices of free sequence ID node in node pool. */
+  std::vector<size_t> free_node_indicess;
+  /*! \brief The map from used paged sequence ID node to its index in node pool. */
+  std::unordered_map<SequenceIDNode*, size_t> used_nodes;
+
+ public:
+  /*! \brief The constructor of sequence Id node pool, allocating memory for each node. */
+  SequenceIDNodePool(size_t num_nodes);
+
+  /*!
+   * \brief Get a radix page from pool, and assign the fields.
+   * \param seq_id The assigned sequence ID of allocated sequence ID node.
+   * \param node The next sequence ID node pointer of allocated sequence ID node.
+   * \return The allocated radix page.
+   * \throw Error if no free radix page available in pool.
+   */
+  SequenceIDNode* Allocate(int64_t seq_id, SequenceIDNode* next);
+
+  /*!
+   * \brief Free a sequence ID node to pool.
+   * \param node The sequence ID node to free.
+   */
+  void Free(SequenceIDNode* node);
+
+  /*! \brief The destructor of sequence Id node pool, freeing memory for each node. */
+  ~SequenceIDNodePool();
+};
+
+/*!
+ * \brief The paged radix tree node data structure.
+ *
+ * The paged radix tree node is similar to original radix tree node, but with the limited length for
+ * prefix in page, so that the memory usage in each page is the same and is fixed once allocated.
+ * Since the page only consists of pointers and int tokens, the page memory layout is int array
+ * indeed. The lower offset is the pointers and page information, while the higher offset is the
+ * stored prefix tokens.
+ *
+ * And since the vocabulary size may be very large, the paged Radix tree is represented
+ * as eft-child, right-sibling binary tree.
+ *
+ * Also, due to possible pop/push front/back tokens in page, the page is designed as circular
+ * buffer, to make full use of each page.
+ *
+ * Each page records the sequence excatly ends with the prefix tokens stored in page. In other word,
+ * all sequences locate in the boundary of each page, or the end of each page.
+ */
+struct RedixPage {
+  /*! \brief The parent page. */
+  RedixPage* parent;
+  /*! \brief The first child page. */
+  RedixPage* first_child;
+  /*! \brief The sibling page shareing the same parent page. */
+  RedixPage* next_sibiling;
+  /*! \brief The head of sequence ID linked list. */
+  SequenceIDNode* seq_ids;
+  /*! \brief The capacity of maximum stored prefix tokens. */
+  size_t capacity;
+  /*! \brief The start offset of stored prefix tokens. The legal value is of [0, capacity). */
+  size_t offset;
+  /*! \brief The length of stored prefix tokens. The legal value is of [0, capacity). */
+  size_t length;
+  /*! \brief The offset of first prefix token in memory layout. */
+  static constexpr int DATA_OFFSET = (sizeof(RedixPage*) * 3 + sizeof(SequenceIDNode*) +
```

Comment:
What about
```
struct RedixPage {
    static const int DATA_OFFSET;
};
inline const int RedixPage::DATA_OFFSET = (sizeof(RedixPage) + sizeof(int32_t) - 1) / sizeof(int32_t);
```

I do not have a strong opinion for that, so it's your call

Reply:
in this case constexpr is better, since const int might relates to a heap variable